File: lib/server/storage.dart
import 'dart:convert';
import 'dart:io';
import 'package:logging/logging.dart';
import 'package:path/path.dart' as p;
import '../core/workspace.dart';
import '../core/module.dart';

final _log = Logger('Storage');

class Storage {
  static String get globalConfigPath =>
      p.join(Platform.environment['HOME']!, '.limousine.json');

  static String pidsDir(String workspacePath) =>
      p.join(p.dirname(workspacePath), '.limousine', 'pids');

  static Future<GlobalConfig> loadGlobalConfig() async {
    final file = File(globalConfigPath);
    if (!await file.exists()) return GlobalConfig();
    return GlobalConfig.fromJson(jsonDecode(await file.readAsString()));
  }

  static Future<void> saveGlobalConfig(GlobalConfig config) async {
    await File(globalConfigPath).writeAsString(
      const JsonEncoder.withIndent('  ').convert(config.toJson()),
    );
  }

  static Future<Workspace> loadWorkspace(String path) async {
    final content = await File(path).readAsString();
    _log.info('Loading workspace from $path');
    try {
      return Workspace.fromJson(jsonDecode(content));
    } on FormatException catch (e, st) {
      _log.severe('Invalid JSON in $path:\n${jsonErrorSnippet(content, e.offset)}', e, st);
      rethrow;
    }
  }

  static Future<void> saveWorkspace(String path, Workspace workspace) async {
    await File(path).writeAsString(
      const JsonEncoder.withIndent('  ').convert(workspace.toJson()),
    );
  }

  static Future<Project?> loadProject(String projectPath) async {
    final projFile = File(p.join(projectPath, 'limousine.proj'));
    if (!await projFile.exists()) return null;
    final content = await projFile.readAsString();
    _log.info('Loading project from ${projFile.path}');
    return Project.fromJson(jsonDecode(content));
  }

  static Future<void> saveProject(String projectPath, Project project) async {
    final projFile = File(p.join(projectPath, 'limousine.proj'));
    await projFile.writeAsString(
      const JsonEncoder.withIndent('  ').convert(project.toJson()),
    );
  }

  /// "Present" = the directory exists AND contains at least one entry other
  /// than `.git`. An empty directory or one that has only a `.git/` left over
  /// from a failed clone counts as *not* present — the UI shows the clone
  /// button, `Git.clone` moves the partial directory aside before retrying.
  static Future<bool> projectExistsOnDisk(
    String rootDir,
    String pathOnDisk,
  ) async {
    final dir = Directory(resolvePath(rootDir, pathOnDisk));
    if (!await dir.exists()) return false;
    await for (final entity in dir.list(followLinks: false)) {
      if (p.basename(entity.path) != '.git') return true;
    }
    return false;
  }

  /// Resolve a workspace-relative `path-on-disk` against [rootDir]. Absolute
  /// paths in the .wksp bypass this and are returned as-is. [rootDir] is
  /// either the parent of the workspace file (default) or `--clone-root` if
  /// the server was started with one.
  static String resolvePath(String rootDir, String pathOnDisk) {
    if (p.isAbsolute(pathOnDisk)) return pathOnDisk;
    return p.normalize(p.join(rootDir, pathOnDisk));
  }

  static String _sanitize(String serviceId) => serviceId.replaceAll('/', '_');
  static String _unsanitize(String filename) => filename.replaceAll('_', '/');

  static Future<void> writePidFile(String workspacePath, String serviceId, int pid) async {
    final dir = Directory(pidsDir(workspacePath));
    if (!await dir.exists()) await dir.create(recursive: true);
    await File(p.join(dir.path, '${_sanitize(serviceId)}.pid')).writeAsString(pid.toString());
  }

  static Future<void> deletePidFile(String workspacePath, String serviceId) async {
    final file = File(p.join(pidsDir(workspacePath), '${_sanitize(serviceId)}.pid'));
    if (await file.exists()) await file.delete();
  }

  static Future<Map<String, int>> loadAllPidFiles(String workspacePath) async {
    final dir = Directory(pidsDir(workspacePath));
    if (!await dir.exists()) return {};
    final result = <String, int>{};
    await for (final entity in dir.list()) {
      if (entity is File && entity.path.endsWith('.pid')) {
        final filename = p.basenameWithoutExtension(entity.path);
        final pid = int.tryParse((await entity.readAsString()).trim());
        if (pid != null) result[_unsanitize(filename)] = pid;
      }
    }
    return result;
  }

  static String jsonErrorSnippet(String content, int? offset) {
    if (offset == null) return content.length > 200 ? '${content.substring(0, 200)}...' : content;
    final lines = content.split('\n');
    var charCount = 0;
    var errorLine = 0;
    for (var i = 0; i < lines.length; i++) {
      charCount += lines[i].length + 1;
      if (charCount > offset) {
        errorLine = i;
        break;
      }
    }
    final start = (errorLine - 3).clamp(0, lines.length);
    final end = (errorLine + 4).clamp(0, lines.length);
    final numbered = <String>[];
    for (var i = start; i < end; i++) {
      final marker = i == errorLine ? '>>>' : '   ';
      numbered.add('$marker ${i + 1} | ${lines[i]}');
    }
    return numbered.join('\n');
  }
}

File: tui/limousine/backend/storage.py
"""File I/O for workspaces, projects, global config, and PID files. Port of
lib/server/storage.dart. Parsing lives in core/models; this does the disk side
plus the numbered JSON-error snippet."""

from __future__ import annotations

import json
import os
from pathlib import Path

from ..core.models import GlobalConfig, Project, Workspace


def global_config_path() -> Path:
    return Path(os.environ["HOME"]) / ".limousine.json"


def pids_dir(workspace_path: str | Path) -> Path:
    return Path(workspace_path).parent / ".limousine" / "pids"


def load_global_config() -> GlobalConfig:
    p = global_config_path()
    if not p.exists():
        return GlobalConfig()
    return GlobalConfig.from_json(json.loads(p.read_text()))


def save_global_config(config: GlobalConfig) -> None:
    global_config_path().write_text(json.dumps(config.to_json(), indent=2))


def load_workspace(path: str | Path) -> Workspace:
    content = Path(path).read_text()
    try:
        return Workspace.from_json(json.loads(content))
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON in {path}:\n{json_error_snippet(content, e.pos)}") from e


def save_workspace(path: str | Path, workspace: Workspace) -> None:
    Path(path).write_text(json.dumps(workspace.to_json(), indent=2))


def load_project(project_path: str | Path) -> Project | None:
    proj_file = Path(project_path) / "limousine.proj"
    if not proj_file.exists():
        return None
    content = proj_file.read_text()
    try:
        return Project.from_json(json.loads(content))
    except json.JSONDecodeError as e:
        raise ValueError(
            f"Invalid JSON in limousine.proj:\n{json_error_snippet(content, e.pos)}"
        ) from e


def save_project(project_path: str | Path, project: Project) -> None:
    (Path(project_path) / "limousine.proj").write_text(json.dumps(project.to_json(), indent=2))


def project_exists_on_disk(root_dir: str | Path, path_on_disk: str) -> bool:
    """Present = directory exists AND has an entry other than `.git`. An empty
    dir or a bare `.git/` from a failed clone counts as not present."""
    d = Path(resolve_path(root_dir, path_on_disk))
    if not d.is_dir():
        return False
    return any(child.name != ".git" for child in d.iterdir())


def resolve_path(root_dir: str | Path, path_on_disk: str) -> str:
    p = Path(path_on_disk)
    if p.is_absolute():
        return str(p)
    return os.path.normpath(os.path.join(str(root_dir), path_on_disk))


def _sanitize(service_id: str) -> str:
    return service_id.replace("/", "_")


def _unsanitize(filename: str) -> str:
    return filename.replace("_", "/")


def write_pid_file(workspace_path: str | Path, service_id: str, pid: int) -> None:
    d = pids_dir(workspace_path)
    d.mkdir(parents=True, exist_ok=True)
    (d / f"{_sanitize(service_id)}.pid").write_text(str(pid))


def delete_pid_file(workspace_path: str | Path, service_id: str) -> None:
    f = pids_dir(workspace_path) / f"{_sanitize(service_id)}.pid"
    if f.exists():
        f.unlink()


def load_all_pid_files(workspace_path: str | Path) -> dict[str, int]:
    d = pids_dir(workspace_path)
    if not d.is_dir():
        return {}
    result: dict[str, int] = {}
    for f in d.glob("*.pid"):
        try:
            result[_unsanitize(f.stem)] = int(f.read_text().strip())
        except (OSError, ValueError):
            continue
    return result


def json_error_snippet(content: str, offset: int | None) -> str:
    if offset is None:
        return content[:200] + "..." if len(content) > 200 else content
    lines = content.split("\n")
    char_count = 0
    error_line = 0
    for i, line in enumerate(lines):
        char_count += len(line) + 1
        if char_count > offset:
            error_line = i
            break
    start = max(0, error_line - 3)
    end = min(len(lines), error_line + 4)
    numbered = []
    for i in range(start, end):
        marker = ">>>" if i == error_line else "   "
        numbered.append(f"{marker} {i + 1} | {lines[i]}")
    return "\n".join(numbered)
